File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.cpp
class Solution {
public:
    int minMutation(string startGene, string endGene, vector<string>& bank) {
         unordered_set<string> bankset(bank.begin(), bank.end());
        
        unordered_set<string> visited;
        queue<string> que;
        que.push(startGene);
        visited.insert(startGene);
        
        int level = 0;
        
        while(!que.empty()) {
            int n = que.size();
            
            while(n--) {
                string curr = que.front();
                que.pop();
                
                if(curr == endGene)
                    return level;
                
                for(char ch : "ACGT") {
                    for(int i = 0; i<curr.length(); i++) {
                        string neighbour = curr;
                        neighbour[i] = ch;
                        
                        if(visited.find(neighbour) == visited.end() && 
                            bankset.find(neighbour) != bankset.end()) {
                            visited.insert(neighbour);
                            que.push(neighbour);
                        }
                    }
                }
            }
            level++;
        }        
        return -1;
    }
};

File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.py
class Solution:
    def minMutation(self, startGene: str, endGene: str, bank: List[str]) -> int:
        bankSet = set(bank)
        visited = set()
        que = []

        que.append(startGene)
        visited.add(startGene)

        level = 0

        while que:
            n = len(que)
            while n:
                n -= 1
                curr = que.pop(0)
                if curr == endGene:
                    return level
                for ch in "ACGT":
                    for i in range(len(curr)):
                        neighbour = curr[:i] + ch + curr[i+1:]
                        if neighbour not in visited and neighbour in bankSet:
                            visited.add(neighbour)
                            que.append(neighbour)
            level += 1
        return -1